### DISCUSSION TITLE
Python wheel files are zip files and can be viewed with VisiData

### DISCUSSION THREAD

**Discussion by frosencrantz** [saulpw/visidata#1539] [General]
For example:
```
vd -n -f zip https://files.pythonhosted.org/packages/d0/e2/dfb2f2a436cda0f0f33dcd4582168010d2536948e90c4549d4b158952116/visidata-2.10.1-py3-none-any.whl
```

This command  uses the builtin VisiData support for unzip-http to allow you to quickly view what files are in a Python Wheel, without having to download the whole file first.

I think this could be combined with @ajkerrigan's plugin for PyPi: https://dev.to/ajkerrigan/data-exploration-quick-wins-with-visidata-13lh/

**Comment by ajkerrigan** [comment]
Yeah it sure can! I'm realizing I didn't mention this shortcut I have in my `.visidatarc` that avoids the need for command line options:

```python
from visidata import vd

vd.open_whl = vd.open_zip                                                                                                                                                   
```

And if you couple it with a command to open a new sheet using the cell value as a path...

```python
from visidata import BaseSheet

BaseSheet.addCommand(' zo', 'open-cell-path', 'vd.push(vd.openSource(cursorCell.value))')
```

Then you have an even handier PyPI remote wheel browser :+1: 

**Edit:** Thanks to @anjakefala's addition in #1514 , the native `zo` keybinding does this. No need for custom shenanigans :).

**Reply by saulpw** [reply]
@ajkerrigan, I'm going to add the same `open_whl` to core visidata.  What else do you have lurking in your visidatarc??

**Reply by ajkerrigan** [reply]
> @ajkerrigan, I'm going to add the same `open_whl` to core visidata. What else do you have lurking in your visidatarc??

Sounds good, sweet :+1: 

> What else do you have lurking in your visidatarc??

My visidatarc is not always a good source of vitamins and minerals. But peeking through I can see some stuff that might be useful beyond my machine. Some might cause clashes and I'm sure they could be written more clearly/sensibly/efficiently. But since you asked...

A couple more loader aliases:

```python
# Terraform state files are JSON, after all...
vd.open_tfstate = vd.open_json

# This is riskier, with `.lock` being very likely to cause conflicts. But I only really feed
# `poetry.lock` files to VisiData so TOML as a default makes sense.
vd.open_lock = getattr(vd, 'open_toml', vd.open_txt)
```

Some stuff for working with IP ranges:

```python
TableSheet.addCommand(None, 'type-ipaddr', 'cursorCol.type=ipaddress.ip_address', 'set type of current column to IP address')
TableSheet.addCommand(None, 'type-ipnet', 'cursorCol.type=ipaddress.ip_network', 'set type of current column to IP network')

@Column.api
def select_supernets(col, addr):
    sheet = col.sheet
    addr = ipaddress.IPv4Network(f'{addr.strip()}{"/32" if "/" not in addr else ""}')
    vd.status(f'selecting rows where {col.name} is a supernet of "{str(addr)}"')
    sheet.select([
        row for row in sheet.rows
        if addr.subnet_of(ipaddress.IPv4Network(col.getValue(row).strip()))
    ])

BaseSheet.addCommand('', 'select-supernets', 'cursorCol.select_supernets(inputExpr("ip or cidr block: "))', 'select rows with CIDR blocks containing a provided IP address or network')
```

A bulk version of `contract-col` to un-expand multiple columns:

```python
def contract_cols(sheet):
    expanded_cols = {c for c in sheet.columns if hasattr(c, 'origCol')}
    for col in expanded_cols:
        col.origCol.width = vd.options.default_width
    sheet.columns = [c for c in sheet.columns if c not in expanded_cols]

BaseSheet.addCommand("g)", "contract-cols", "contract_cols(sheet)")
```

Some helpers for dealing with files that have mixed/nested data types:

```python
def openCellAs(col, row, filetype='txt'):
    with NamedTemporaryFile('w+') as ntf:
        ntf.write(str(col.getValue(row)))
        ntf.seek(0)
        vd.push(vd.openPath(vd.Path(ntf.name), filetype=filetype))

@TableSheet.api
def parseJsonColumn(sheet, col):
    newcol = ExprColumn("json.loads(curcol)", curcol=col)
    newcol.name = f"{col.name}_json"
    sheet.addColumnAtCursor(newcol)

BaseSheet.addCommand("zo", "open-cell-as", "openCellAs(cursorCol, cursorRow, filetype=vd.input('open cell with loader: ', 'filetype'))")
BaseSheet.addCommand(" zj", "open-cell-json", "openCellAs(cursorCol, cursorRow, filetype='json')")
BaseSheet.addCommand(" j", "addcol-json", "sheet.parseJsonColumn(cursorCol)")
```

^^ I need to rename `zo`, since @anjakefala made VisiData more useful out of the box. The nerve!

**Reply by frosencrantz** [reply]
Thanks for sharing @ajkerrigan!  I wouldn't mind all of these being added to base VisiData!

And I would add one more format `har` files are also json.  They have more structure, but that is a starting point.

**Reply by geekscrapy** [reply]
@ajkerrigan thanks for these snippets! Maybe we could get the IP commands baked in as they are stdlib stuff!

Also love the `open-cell-as`, maybe this could also be baked in and assigned `gzo`?

**Reply by saulpw** [reply]
Hm, should `zo` be `open-cell-as` or `open-cell-file`?  What are you people putting in your cells :)

**Reply by geekscrapy** [reply]
This is v. true @saulpw ! I often have links (when loaded a table from an html page), or embedded json, or even a filename from `grep -h` output! so for me, I wouldn't consistently need one over another. Maybe it should be `open-cell-as`. I see it more of a powertool anyway so someone should know more about what they are doing than a casual user that opens a flat dataset?

**Reply by frosencrantz** [reply]
I would use both also.

**Comment by frosencrantz** [comment]
There are a few other file formats that are based on zip format according to [Wikipedia]:
- jar
  - war
  - ear
  - rar
- Office Open XML 
  - docx
  - docm
- odf (Open Document)
- xpi (Mozilla Extension)
- .lottie files

Also, `.har` files are JSON based.

It is not perfect showing these files as their base types, but it is more useful than if VisiData completely ignored the format.

@saulpw what do you think?


[Wikipedia]:   https://en.wikipedia.org/wiki/ZIP_(file_format)

**Reply by saulpw** [reply]
Hmm...there are many dozens or hundreds of these formats, I'm guessing.  What if we included it in formats.jsonl, and then generated this list from that?

**Reply by frosencrantz** [reply]
@saulpw would the generated list be installed with VisiData?  That is reasonable.  Even though they are zip or json formats, they have additional structure.  It would be great if VisiData knew the additional structure, but having only the base format is better than nothing.
